File: workshop-exercises/02-clean-code/go/application/paypal_processor.go
package application

import (
	"fmt"
	"time"
)

// =============================================================================
// PAYPAL PROCESSOR
// =============================================================================

type PayPalProcessor struct {
	feePercent float64
}

func NewPayPalProcessor() PaymentProcessorInterface {
	return &PayPalProcessor{
		feePercent: 3.49,
	}
}

func (p *PayPalProcessor) ProcessPayment(amount float64) (string, error) {
	return p.executePaymentProcessing(amount)
}

func (p *PayPalProcessor) executePaymentProcessing(amount float64) (string, error) {
	fee := p.calculateProcessingFee(amount)
	total := p.calculateTotalAmount(amount, fee)
	p.simulateProcessingDelay()
	return p.formatPaymentResult(total, fee), nil
}

func (p *PayPalProcessor) calculateProcessingFee(amount float64) float64 {
	return p.computeFeeFromAmount(amount)
}

func (p *PayPalProcessor) computeFeeFromAmount(amount float64) float64 {
	return amount * p.feePercent / 100
}

func (p *PayPalProcessor) calculateTotalAmount(amount float64, fee float64) float64 {
	return p.addFeeToAmount(amount, fee)
}

func (p *PayPalProcessor) addFeeToAmount(amount float64, fee float64) float64 {
	return amount + fee
}

func (p *PayPalProcessor) simulateProcessingDelay() {
	p.waitForProcessing(150 * time.Millisecond)
}

func (p *PayPalProcessor) waitForProcessing(duration time.Duration) {
	time.Sleep(duration)
}

func (p *PayPalProcessor) formatPaymentResult(total float64, fee float64) string {
	return fmt.Sprintf("PayPal: $%.2f (fee: $%.2f)", total, fee)
}

File: workshop-exercises/02-clean-code/csharp/Application/Services/PayPalProcessor.cs
using Workshop.Application.Interfaces;

namespace Workshop.Application.Services
{
    public class PayPalProcessor : IPaymentProcessor
    {
        private readonly double _feePercent = 3.49;

        public string ProcessPayment(double amount)
        {
            return ExecutePaymentProcessing(amount);
        }

        private string ExecutePaymentProcessing(double amount)
        {
            var fee = CalculateProcessingFee(amount);
            var total = CalculateTotalAmount(amount, fee);
            SimulateProcessingDelay();
            return FormatPaymentResult(total, fee);
        }

        private double CalculateProcessingFee(double amount)
        {
            return ComputeFeeFromAmount(amount);
        }

        private double ComputeFeeFromAmount(double amount)
        {
            return amount * _feePercent / 100;
        }

        private double CalculateTotalAmount(double amount, double fee)
        {
            return AddFeeToAmount(amount, fee);
        }

        private double AddFeeToAmount(double amount, double fee)
        {
            return amount + fee;
        }

        private void SimulateProcessingDelay()
        {
            WaitForProcessing(150);
        }

        private void WaitForProcessing(int milliseconds)
        {
            Thread.Sleep(milliseconds);
        }

        private string FormatPaymentResult(double total, double fee)
        {
            return $"PayPal: ${total:F2} (fee: ${fee:F2})";
        }
    }
}